src/move_and_slide.rs

```diff
@@ -99,3 +100,82 @@ pub fn move_and_slide(
         }
     }
 }
+
+fn similar_plane(a: Vec3, b: Vec3) -> bool {
+    a.dot(b) > SIMILARITY_THRESHOLD
+}
+
+fn solve_collision_planes(
+    velocity: Vec3,
+    hits: &[Vec3],
+    original_velocity_direction: Vec3,
+) -> Vec3 {
+    // Early out if we have no velocity or no hits
+    if velocity.length_squared() <= 0.0 || original_velocity_direction.length_squared() <= 0.0 {
+        return Vec3::ZERO;
+    }
+
+    if hits.is_empty() {
+        return velocity;
+    }
+
+    // Do our initial rejection to calculate the sliding velocity.
+    let first_hit_normal = hits[hits.len() - 1];
+    let velocity_direction = velocity.normalize_or_zero();
+    if velocity_direction.dot(first_hit_normal) >= 0.0 {
+        return velocity;
+    }
+    let initial_velocity = velocity.reject_from(first_hit_normal);
```

Comment:
We can just use `velocity` directly here, right? The magnitude doesn't matter for the dot product's sign here, only the direction does
```suggestion
    // Do our initial rejection to calculate the sliding velocity.
    let first_hit_normal = hits[hits.len() - 1];
    if velocity.dot(first_hit_normal) >= 0.0 {
        return velocity;
    }
    let initial_velocity = velocity.reject_from(first_hit_normal);
```